library/x509_crt.c

```diff
@@ -3527,9 +3528,16 @@ static int x509_crt_verify_chain(
         }
 
         /* signature was checked while searching parent */
-        if( signature_is_good != X509_SIGNATURE_IS_GOOD )
+        signature_is_good_fi = signature_is_good;
+        if( signature_is_good_fi != X509_SIGNATURE_IS_GOOD )
```

Comment:
Instead of ensuring that if the signature is not good, would it be better to use the "default flow assumes failure"? In this case, the condition is checked twice and the flag is inserted twice, so it is good as it stands.

Reply:
I thought about it but couldn't come up with a nice way of implementing that here. My initial idea was to first set the flag, then unset it if the signature is good, but that doesn't quite work because we'd risk clearing too much. If you have other ideas, feel free to share :)

Reply:
Oh, I replied from the github notifications so didn't have the context and failed to notice that you were suggesting the same idea, so let me expand on why I don't think that will work: once we introduce `X509_BADCERT_FI_EXTRA` we need extra logic to make sure that we only clear that part if it wasn't already set before, but then this extra logic also needs protecting against faults...

Reply:
Btw, how does the glitching actually work in this case. Could the attacker glitch it so that this check fails and setting the `MBEDTLS_X509_BADCERT_NOT_TRUSTED` doesn't get set since it just passes the whole if-clause?

Reply:
"For example, when checking authorization, the critical condition is a query if the authorization was successful. There is no reason to double check if the authorization failed."

Reply:
Damn, you're right, I got the nesting wrong. Fixed it in a new commit (hopefully). Nice catch, thanks!